api/lib/domain/services/obfuscation-service.js

```diff
@@ -0,0 +1,44 @@
+const _ = require('lodash');
+
+const CONNEXION_TYPES = {
+  username: 'username',
+  email: 'email',
+  samlId: 'samlId',
+};
+const ASTERISK_OBFUSCATION = '***';
+const USERNAME_SEPARATOR = '.';
+const EMAIL_SEPARATOR = '@';
+
+const TWO_PARTS = 2;
+
+function getUserAuthenticationMethodWithObfuscation(user) {
+  if (user.samlId) return { authenticatedBy: CONNEXION_TYPES.samlId, value: null };
+
+  if (user.username) {
+    const username = usernameObfuscation(user.username);
+    return { authenticatedBy: CONNEXION_TYPES.username, value: username };
+  }
+  if (user.email) {
+    const email = emailObfuscation(user.email);
+    return { authenticatedBy: CONNEXION_TYPES.email, value: email };
+  }
+}
+
+function emailObfuscation(email) {
+  const obfuscationValue = ASTERISK_OBFUSCATION.concat(EMAIL_SEPARATOR);
+  const parts = _.split(email, EMAIL_SEPARATOR, TWO_PARTS);
+  return email.charAt(0).concat(obfuscationValue).concat(parts[1]);
+}
+
+function usernameObfuscation(username) {
+  const obfuscationValue = ASTERISK_OBFUSCATION.concat(USERNAME_SEPARATOR);
+  const parts = _.split(username, USERNAME_SEPARATOR, TWO_PARTS);
+  const indexOfLastChar = parts[1].length - 1;
+  return username.charAt(0).concat(obfuscationValue).concat(parts[1].charAt(0)).concat(ASTERISK_OBFUSCATION).concat(parts[1].charAt(indexOfLastChar));
```

Comment:
Pour alléger la lecture, on peut utiliser les [template strings](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Litt%C3%A9raux_gabarits).
On aurait ainsi le code suivant :

```suggestion
  const parts = _.split(email, EMAIL_SEPARATOR, TWO_PARTS);
  return `${_.first(email)}${ASTERISK_OBFUSCATION}${EMAIL_SEPARATOR}${_.last(parts)}`;
}

function usernameObfuscation(username) {
  const parts = _.split(username, USERNAME_SEPARATOR, TWO_PARTS);
  const name = _.last(parts);
  const indexOfLastChar = parts[1].length - 1;
  const usernameFirstChar = username.charAt(0);
  return `${_.first(username)}${ASTERISK_OBFUSCATION}${USERNAME_SEPARATOR}${_.first(name)}${ASTERISK_OBFUSCATION}${_.last(name)}`;
```

Reply:
Thanks, c'est plus lisible et les tests passent :)